동기화 메모장 [STEP 2] Nicholas, 신나, yeha

smart8612 opened the pull request:
안녕하세요, 콘 @protocorn93 ! 

예하 @ye-ha 니코 @Kim-EunsooSilver 신나 @smart8612 팀입니다!
스텝 2 PR 보냅니다. 이번 PR도 잘 부탁드립니다!🙏
저희 스텝2에서 1. 스와이프했을 때 공유와 삭제할 수 있는 기능 2. 삭제했을 때 오른쪽 디테일 뷰에서도 사라지는 기능은 아직 구현하지 못해서 이 부분은 피드백 반영하면서 구현해보겠습니다.

# 고민했던 점

## 프로젝트 설계 구현 패턴 (MVVM)

### 기존의 `NoteModelManager`를 `NoteModel`과 `NoteViewModel`로 분리 구현하였습니다.

- Model 객체에 View Presenting 로직을 함께 구현하였을 때 다음과 같은 문제점을 갖습니다.
    - Model에 B 화면을 위한 ViewController에서 보여줄 데이터 변환 API가 담겨있다고 가정합니다.
    - A 화면을 위한 ViewController에서 이 Model을 재사용하고 싶습니다.
    - A 화면의 ViewController 입장에서는 Model에 불필요한 API 가 많이 보이게 됩니다.
    - A 화면을 위한 데이터 변환 로직도 Model에 구현한다면, Model을 사용하는 View의 수가 증가함에 따라 Model에 구현되는 데이터 변환 API가 증가하게 됩니다.
    - 즉, Model 객체를 확장하거나 재사용하는 부분에 있어서 손해라고 생각했습니다.
- 따라서 ViewModel 객체를 도입하여 각 화면에 알맞게 데이터를 변환하는 로직을 Model로 부터 변환시켜주었습니다.

### 프로젝트 전체에서 데이터가 단방향 흐름을 갖도록 설계하였습니다.

- UI가 상태(State)값을 갖는 기존의 MVC 구조에서는 다음의 문제점이 발생합니다.
    - 데이터가 갱신되면 Controller 객체가 View 와 데이터 사이에서 동기화 작업을 수행해줘야 합니다.
    - 다시 말해, 관리 해줘야 하는 데이터가 많기 때문에 데이터의 양에 비례하여 Controller 객체 구현이 복잡해질 수 있습니다.
- 따라서 데이터의 근원을 Model 객체 하나로 통합 관리해주고, Model 데이터의 상태로만 View 가 갱신되는 구조를 만들고 싶었습니다.

### Model 데이터의 상태를 View에 Binding 하기 위해 사용한 방법은 다음과 같습니다.

1. CoreData 데이터 바인딩을 위해 `NSFetchedResultsController` 사용하였습니다.
    - 일반 값타입 인스턴스의 경우 Property Observer를 통해 데이터의 변동을 손쉽게 감지할 수 있습니다.
    - CoreData를 사용하는 경우 데이터의 CRUD 작업을 Controller의 ViewContext 로만 진행합니다.
    - 따라서 데이터베이스 상에 데이터 변화를 감지할 방법에 관해 고민하게 되었습니다.
    - `NSFetchedResultsController`는 CoreData가 제공하는 View Syncronization 도구로, 특정 Request의 데이터를 관리해줍니다.
    - 특히, CoreData 저장소의 데이터가 변동되면 controller 의 delegate 에게 이벤트를 전달해줍니다.
    - delegate 이벤트를 통해 View를 갱신시켜주는 코드를 실행시킬 수 있었습니다.
2. 데이터에 기반하여 TableView를 갱신하기 위해 `UIDiffableDataSource` 객체를 사용하였습니다.
    - 기존의 UITableView 메서드를 통해 특정 행의 데이터를 CRUD 하는 방식은 UI가 데이터 상태값을 갖게되어 관리할 데이터가 늘어난다는 단점이 있었습니다.
    - 데이터를 제공하여 데이터 혹은 그 변동 사항에 기반한 UI 갱신 코드를 구현하기 위해 `UITableViewDiffableDataSource` 클래스를 사용하게 되었습니다.
    - Core Data 로 부터 받아온 데이터를 snapshot 으로 가공하여 데이터 소스 객체를 통해 apply 해주는 단순한 사용법으로 데이터의 근원을 줄일 수 있었습니다.

### `UIViewController`의 새로운 관점

- 데이터 처리 및 앱의 비즈니스 로직이 Model과 ViewModel에 의해 분리 구현되면서 ViewController는 주로 화면을 그려주고 UI를 갱신하는 코드가 잔류하게 되었습니다.
- 따라서 기존의 Controller의 역할 보다는 앱 상에서 View의 역할에 좀더 가깝게 구현되었습니다.
- 다만, 기존과 차이점이라면 시스템 이벤트와 사용자의 터치 이벤트 등을 제어할 수 있는 View 라는 관점으로 보게 된 것 같습니다.

### Model 과 ViewController 간의 강한 결합성이 잔류하고 있습니다.

- 기존 프로젝트 구조에서 객체간 역할 분리를 세분화 하던 중 프로토콜을 통한 객체간의 소통 인터페이스 분리 구현을 하지 못했습니다.
- 다음 커밋에서 시간이 남는다면 추상 프로토콜로 인터페이스 구현하는 것을 목표로 하고 있습니다.

---

## Alert 관련 재사용 템플릿 구현 → 클로저 사용

- 더보기 버튼이 눌렸을 때, action Sheet 스타일의 `UIAlertController` 가 우선 띄워집니다.
- 해당 alert controller 에는 Share, Delete Action이 있습니다.
- 이 때 Share를 누르게 되면 `UIActivityViewController` 가 띄워집니다.
- Delete를 누르게 되면 삭제를 다시 한 번 확인하는 `UIAlertController`가 띄워집니다.
- 이 과정에서 아래 `detailBarButtonItemDidTap` 메서드가 길어지고, 같은 작업을 상황이 두 번 이상 발생하였습니다.
    
    ```swift
        @objc
        private func detailBarButtonItemDidTap() {
            guard let identifier = identifier else {
                return
            }
            let shareAction = UIAlertAction(title: "Share", style: .default) { [self] _ in
                presentActivityView(items: [noteDetailTextView.text ?? ""]) { activityViewController in
                    activityViewController.popoverPresentationController?.barButtonItem = detailBarButtonItem
                }
            }
            let deleteAction = UIAlertAction(title: "Delete", style: .destructive) { _ in
                self.presentAlert(title: "진짜요?", message: "정말로 지워요?") { alert in
                    let actions = [
                        UIAlertAction(title: "취소", style: .cancel),
                        UIAlertAction(title: "삭제", style: .destructive) { _ in
                            self.viewModel.deleteNote(identifier: identifier)
                        }
                    ]
                    alert.addAction(actions)
                }
            }
            presentAlert(preferredStyle: .actionSheet) { alertController in
                alertController.modalPresentationStyle = .popover
                alertController.popoverPresentationController?.barButtonItem = detailBarButtonItem
                alertController.addAction([shareAction, deleteAction])
            }
        }
    ```
    
- 그래서 재사용할 수 있는 템플릿을 만들고, 클로저를 통해 사용하도록 했습니다.
- `UIActivityViewController`는 한 번 띄워지지만, 가독성을 위해 함께 생성했습니다.
    
    ```swift
    extension UIViewController {
        
        func presentAlert(
            title: String? = nil,
            message: String? = nil,
            preferredStyle: UIAlertController.Style = .alert,
            handler: ((UIAlertController) -> Void)
        ) {
            let alertController = UIAlertController(title: title, message: message, preferredStyle: preferredStyle)
            handler(alertController)
            self.present(alertController, animated: true)
        }
        
        func presentActivityView(
            items: [Any],
            activities: [UIActivity]? = nil,
            handler: ((UIActivityViewController) -> Void)
        ) {
            let activityViewController = UIActivityViewController(activityItems: items, applicationActivities: activities)
            handler(activityViewController)
            self.present(activityViewController, animated: true)
        }
        
    }
    
    extension UIAlertController {
        
        func addAction(_ action: [UIAlertAction]) {
            action.forEach { self.addAction($0) }
        }
        
    }
    ```
    

---

---

# 해결이 되지 않은 점

## `textViewDidChange()` 관련

- `textViewDidEndEditing` 메서드를 사용했을 때, end되는 시점을 잡기가 어려워 `textViewDidChange`메서드에서 텍스트 수정에 따른 자동 저장을 하고 있습니다.
- 해당 메서드의 작동을 확인한 결과 이 메서드가 지나치게 자주 호출되어 앱이 원활하게 작동하지 않습니다.
    
    ```swift
        func textViewDidChange(_ textView: UITextView) {
            guard let identifier = identifier else {
                return
            }
            
            let text = textView.text ?? ""
            let result = text.split(separator: "\n", maxSplits: 1).map { String($0) }
            
            let title = result[0]
            let body = result.count >= 2 ? result[1] : ""
            
            viewModel.updateNote(identifier: identifier, title: title, body: body)
        }
    ```
    

---

## Cancel 액션 구현 관련

![Screen Shot 2022-02-21 at 9 40 39 PM](https://user-images.githubusercontent.com/25794814/154957128-8ec06e4e-7aa5-4cdd-b943-fcb1a5b5aef6.png)

- 프로젝트 명세에는 액션시트에 `Cancel 액션`이 구현되어 있습니다.
- 프로젝트 도중에 캔슬 액션을 코드상에 추가하였으나, 시뮬레이터에서는 노출이 되지 않았습니다.
- 여러 레퍼런스에서 iPadOS에서는 `액션시트 이외의 영역을 터치` 함으로서 대체될수 있고, 취소 버튼을 포함하게 될 경우 사용자에게 혼란을 야기할 수 있다고 한다는 내용을 보았습니다.
- 하지만 프로젝트 명세에 Cancel 액션이 있어, 방법을 찾아 보았지만 찾을 수 없어 구현하지 못했습니다.
- 캔슬 버튼이 노출될 수 있는 방법이 있을까요? 혹은 현재 상태도 괜찮을까요?

---

## UIActivityViewController의 Options 버튼 관련

![Screen Shot 2022-02-21 at 9 41 21 PM](https://user-images.githubusercontent.com/25794814/154957211-e95c5473-e276-4ea4-b98c-5626ea57163d.png)

- 프로젝트 명세에 있는 `Options >` 버튼 관련하여 관련된 정보 및 구현 방법을 찾을 수가 없었습니다.
- 아이폰에서 확인해본 결과 safari에서 이 버튼을 확인할 수 있었습니다.
- 해당 버튼의 기능과 구현 방법을 알 수 있을까요?

---

---

# 조언을 얻고 싶은 부분

## lazy var 관련

- `NoteDetailViewController`의 `detailBarButtonItem` 프로퍼티의 선언을 상수로 구현하고, 액션을 연결하는 도중에 버튼의 액션이 제대로 동작되지 않는 이슈가 발생하였습니다.
- 앱을 구동했을 때, 아래 네모의 더보기 버튼을 누르면 아무 동작이 하지 않는 이슈입니다.


![Simulator Screen Shot - iPad Pro (12 9-inch) (4th generation) - 2022-02-21 at 20 53 29](https://user-images.githubusercontent.com/60090790/154951095-444581b6-cbb0-486d-a82e-83b865efa597.png)


- 왼쪽의 노트 테이블 리스트의 항목 중 하나를 눌러서 노트의 상세 페이지를 들어간 후 `TextView`를 클릭해서 수정하는 상태가 되어야 더보기 버튼이 동작하는 상황이었습니다.
- `lazy var` 로 변경을 하였고 제대로 동작이 되었습니다.
    
    ```swift
    private lazy var detailBarButtonItem: UIBarButtonItem = {
        let image = UIImage(systemName: "ellipsis.circle")
        let barButtonItem = UIBarButtonItem(
            image: image,
            style: .plain,
            target: self,
            action: #selector(detailBarButtonItemDidTap))
        return barButtonItem
    }()
    ```
    
    - 현재는 노트가 뜨지 않는 첫 화면에서 버튼이 눌려도 동작하지 않도록 지정해둔 상태입니다.
- 문제점을 분석하는 과정
    1. responder의 측면
        1. `UISplitViewController`는 두개 이상의 child view controller 를 동시에 띄어주는 container view controller 이다.
        2. responder chain 의 측면에서 터치 이벤트가 `NoteDetailViewController`(Secondary ViewController)가 활성화되기 전까지는 `NoteListTableViewController`(Primary ViewController) 로만 전달되어서 bar button item 클릭시 action 메서드가 호출되지 않았을 것이다.
    2. target 측면
        1. bar buttom item 의 target 으로 등록된 `NoteDetailViewController` 인스턴스가 `UISplitViewController`에서 활성화되기 전에 target 이 등록되어 문제가 발생했을 것이다.
        2. Lazy 키워드를 사용하여 `NoteDetailViewController` 인스턴스 생성 후 필요할 때 bar button item 을 생성하여 target을 등록하면 target action이 잘 동작할 것이다.
        3. 실제로 해보니 잘 동작..
- 저희가 생각한 부분이 맞는지 더 생각해볼 부분이 있는지 궁금합니다.

---

## 데이터 바인딩 역할 분리 관련

- CoreData 의 데이터가 변경되면 `NSFetchedResultController`가 모델의 변화를 감지하여 `NSFetchedResultsControllerDelegate`를 통해 이벤트 전달하는 구조를 갖습니다.
    - ViewController, ViewModel, Model 중 어떤 객체가 `NSFetchedResultController` 의 delegate 역할을 담당하면 좋을지 고민해보았습니다.
        1. 만약 ViewModel이 delegate 역할을 담당한다면, ViewModel 클래스에 `NSObject`를 상속해줘야 `NSFetchedResultsControllerDelegate` 프로토콜을 채택할 수 있습니다.
            - 데이터의 관문이라는 관점에서 ViewModel을 보았을 때 `NSObject`를 상속받는 것이 적절해보이지 않습니다.
        2. 만약 ViewController가 delegate 역할을 담당한다면, ViewController에서 Model을 참조할 수 있도록 private 을 풀어주거나, `NSFetchedResultController`의 인스턴스가 ViewModel로 옮겨와야 합니다.
            - 다만, ViewModel 은 Model 데이터의 문지기이고 ViewController는 UI 제어 역할을 담당하기에 데이터를 제어하는 Controller 코드가 들어가는 것이 좋을것으로 생각하고 있습니다.
    - 선택의 딜레마를 어떤 방식으로 풀어가볼 수 있을까요?

---

## CoreData, Foundation 관련

- `import CoreData` 하게 되면 `CoreData`가 내부적으로 `Foundation`을 참조하고 있습니다.
- 이 둘을 개별적으로 import 해주는 것이 좋을까요?
- 더 나아가서 프레임워크 코드를 import 하는 것이 다른 개발자에게 어떤 인상을 줄 수 있을까요?
- 현업에서는 어떻게 사용되고 있고, 콘은 어떻게 하시는지 궁금합니다!

Kim-EunsooSilver committed:
refactor: diffableDataSource 구현 #5

- NoteListTableViewDiffableDataSource 생성
- Section 생성

hayeonhee committed:
fix: iOS 14 이전 기기에서 크래시나는 버그 수정 #4

- view hierarcy에 cell이 생성되기 전에 animating 효과를 켜서 크래시나는 버그 발견
- 해당 작업을 cell이 register된 후에 apply하면 해결!

smart8612 committed:
feat: ViewController와 Model 사이 클로저를 통한 통신 구현 #5

smart8612 committed:
feat: Core Data Model Entity 구현 #5

hayeonhee committed:
feat: PersistentDataManager 생성 #5

Kim-EunsooSilver committed:
refactor: Coredata이용한 메모 저장 및 읽기 기능 구현 #5

smart8612 committed:
refactor: PersistentDataManager create, fetch 메서드 수정 #5

- create 메서드 handler를 통한 managedObject 설정
- 에러를 외부로 전파시키도록 수정

hayeonhee committed:
test: create, fetch, update메서드 테스트케이스 수정 및 추가 #5

hayeonhee committed:
feat: PersistentDataManager 내 update 메서드 구현 #5

- NoteModelManager 내 updateNote 메서드 구현

hayeonhee committed:
feat: CDNote Class 엔티티 구현 #5

- 자동 생성 기능 동작 오류로 CodeGen옵션을 Class Definition -> Manual로 변경
- fetchNoteRequest(), fetchNoteRequest(with identifier: UUID) 메서드 구현

hayeonhee committed:
feat: PersistentDataManager 내 delete 메서드 구현 #5

hayeonhee committed:
test: 테스트케이스 내 수정된 메서드 적용 #5

Kim-EunsooSilver committed:
feat: NSFetchedResultsController 구현 및 데이터 바인딩 #5

- fetchData 구현
- NSFetchedResultsController의 delegate 지정
- noteData 컨버팅

smart8612 committed:
feat: add barbuttonitem 클릭시 노트 생성 기능 구현 #5 #6

smart8612 committed:
refactor: Model데이터와 View로직 분리 및 저장기능 구현 #5 #6

- ViewModel과 Model에 CRUD API 구현
- textView 변경시 노트 수정 내용 자동 저장 기능 구현
- noteModel 프로토콜 제거

hayeonhee committed:
feat: action sheet 및 activity view 구현 #6

- 더보기 버튼에 action sheet 스타일의 UIAlertController 연결
- UIAlertController 내에 공유, 삭제 기능 추가
- 공유 누르면 UIActivityViewController에 데이터 전달
- 삭제 누르면 노트 삭제

Kim-EunsooSilver committed:
feat: delete 확인 alert 구현 #6

- 재사용성을 고려하여 UIViewController, UIAlertController extension 구현

smart8612 committed:
style: 개행 및 불필요한 코드 수정 #6

smart8612 committed:
refactor: 접근제어자 및 코드 개행 수정 #6

smart8612 committed:
fix: fetchSortedNoteRequest 정렬 내림차순 수정 #6

hayeonhee committed:
test: 타겟 멤버십 변경 및 테스트케이스 삭제 #5

hayeonhee committed:
style: 파일명 및 파일 순서 수정 #5

Kim-EunsooSilver committed:
feat: 노트 제거 시 디테일 뷰에서 삭제처리 구현 #6

smart8612 committed:
feat: CRUD 시에 선택된 Cell 동기화 기능 구현 #6

smart8612 commented:
# 추가적인 업데이트 사항

TableView Cell 에서 Swipe 하여 삭제하는 기능과 선택된 Cell 에 맞추어 Detail TextView를 갱신하는 기능을 구현하였습니다.

다만, swipe를 통한 delete 기능을 사용했을 때 선택된 Cell 이 회색 배경으로 강조되는 효과가 적용되지 않는 문제점에 봉착하였습니다.

## 문제 동작
![Screen Recording 2022-02-22 at 4 04 50 PM](https://user-images.githubusercontent.com/25794814/155080539-4a29a89b-1c16-4b1f-b54d-9fccd0fc5b15.gif)

protocorn93 commented on CloudNotes/CloudNotes/Controller/NoteListTableViewController.swift:
```diff
@@ -1,64 +1,109 @@
+final class NoteListTableViewController: UITableViewController {
     
-    private var noteModelManager: NoteModel
+    private var viewModel: NoteViewModel
```
`let`으로 정의되어도 무방해보네요~!

hayeonhee replied:
이 부분 `let`으로 수정했습니다!

protocorn93 commented on CloudNotes/CloudNotes/Controller/NoteListTableViewController.swift:
```diff
@@ -1,64 +1,109 @@
+                    )
+                }
+                
+                return cell
```
`NoteListTableViewDiffableDataSource`내부에서 `DiffableDataSource`를 생성할 때 함께 넘기는 클로저는 옵셔널을 반환을 허용하니까 애초에 셀을 dequeue할 때 `guard-let`으로 언랩핑하고 `return nil`을 해주는 것이 전체적으로 좀 더 깔끔한 코드 구조일 수 있지 않을까 하는 생각입니다!

smart8612 replied:
* cell에 대한 다운캐스팅 작업을 guard 구문을 통해 처리하도록 수정하였습니다.
* 명시적으로 nil 을 반환하도록 수정하였습니다!
* 클로저 반환 타입과 코드의 의미를 좀더 명확하게 이해할 수 있게된 것 같습니다.

protocorn93 commented on CloudNotes/CloudNotes/Controller/NoteListTableViewController.swift:
```diff
@@ -1,64 +1,109 @@
+    
+    private func updateTable(_ type: NSFetchedResultsChangeType) {
+        
+        print(type.rawValue)
```
불필요한 출력문으로 제거해도 무방해보입니다!

hayeonhee replied:
아이고😖 삭제했습니다!

protocorn93 commented on CloudNotes/CloudNotes/Controller/NoteListTableViewDiffableDataSource.swift:
```diff
@@ -0,0 +1,36 @@
+
+final class NoteListTableViewDiffableDataSource: UITableViewDiffableDataSource<Section, Note> {
+    
+    private var viewModel: NoteViewModel
```
이 프로퍼티도 `let`으로 정의되어도 무방해보입니다.

hayeonhee replied:
맞습니다. `let` 으로 바꿔주었습니다!

protocorn93 commented on CloudNotes/CloudNotes/Model/Section.swift:
```diff
@@ -0,0 +1,7 @@
+import Foundation
+
+enum Section {
+    
+    case main
+    
+}
```
DiffableDataSource에서 사용하는 `Section`, `Item`과 같은 타입은 해당 뷰컨트롤러 내부에서 정의해두는 것이 좋아요. 

뷰를 그리는데 표현하는 모델이기 때문에 그리고 뷰컨트롤러 내부에서 다른 파일로 점프하지 않고 볼 수 있다는 장점 때문에 저는 이런 타입은 뷰컨트롤러 내부에 정의하곤합니다.

smart8612 replied:
Section 타입은 특정 ViewController에서 사용하는 태이블 뷰의 섹션을 구문한다는 측면에서 콘에 생각에 동감합니다.
해당 코드를 ViewController 내부로 옮김으로써 해당 타입에 접근할 때 어떤 화면에서 사용되는지 네임스페이스를 통해 명시적으로 알 수 있도록 개선되었습니다.

protocorn93 commented on CloudNotes/CloudNotes/Model/PersistentDataManager.swift:
```diff
@@ -0,0 +1,68 @@
+import Foundation
+import CoreData
+
+final class PersistentDataManager {
+    
+    // MARK: - Core Data stack
+    lazy var persistentContainer: NSPersistentCloudKitContainer = {
```
이 프로퍼티는 `private` 접근제어자가 필요해보입니다.

hayeonhee replied:
`private` 접근제어자 지정했습니다.

protocorn93 commented on CloudNotes/CloudNotes/Model/NoteViewModel.swift:
```diff
@@ -0,0 +1,72 @@
+    
+    private let model: NoteModel
+    var updatePrimaryHandler: (() -> Void)?
+    var updateSecondaryHandler: ((NSFetchedResultsChangeType) -> Void)?
```
이 두 클로저의 네이밍이 좀 더 명확한 의미를 전달할 수 있는 네이밍이면 좋을 것 같아요.

protocorn93 replied:
그리고 둘로 분리해놓은 이유가 궁금합니다.

smart8612 replied:
* `updatePrimaryHandler`는 데이터가 변경되었을 때 diffableDataSource가 테이블 뷰에 스냅샷을 apply 해주는 코드입니다.
* `updateSecondaryHandler`는 데이터가 변경되었을 때 데이터의 변동 타입 (생성, 삭제, 이동, 수정) 상황에 따라 대응하여 UI를 갱신해주는 코드입니다.
* 코드를 구현할 때 프로토타입 코드 형태로 위와 같이 네이밍 하였지만 클로저의 각 역할에 따라서 다음과 같이 네이밍을 변경하였습니다.
    * `updatePrimaryHandler` -> `updateUIHandler`
    * `updateSecondaryHandler`-> `updateUIByDataHandler`

smart8612 replied:
* 두 코드가 분리 된 이유는 데이터가 변동되었다는 이벤트에서 더 나아가 데이터의 변동 세부 사항에 따라 다르게 UI를 갱신해줄 필요가 있었기 때문입니다. 
* 대표적으로, 노트가 삭제되면 첫번째 행을 자동으로 선택해주는 등의 기능 예시가 있습니다!

protocorn93 commented on CloudNotes/CloudNotes/Controller/NoteDetailViewController.swift:
```diff
@@ -1,6 +1,16 @@
+            noteDetailTextView.isEditable = identifier != nil
+            if identifier == nil {
+                noteDetailTextView.text = ""
+            }
```
이 4줄의 코드가 어떤걸 의미하는지 설명해주실 수 있으신가요? 의미 전달이 조금 모호한것같네요!

smart8612 replied:
DetailViewController가 보여줘야 할 노트가 있을 때에는 identifier가 UUID 값을 갖고 있으며 없을시에는 nil 값을 갖고 있습니다.
따라서 보여줘야 할 노트가 없을때 텍스트 뷰를 비워주고 편집기능을 비활성화하도록 구현한 코드입니다.
함수 분리를 통해 코드의 의미가 명확하게 전달될 수 있도록 개선시켰습니다.

protocorn93 commented on CloudNotes/CloudNotes/Controller/NoteListTableViewController.swift:
```diff
@@ -1,64 +1,109 @@
-    init(model: NoteModel) {
-        self.noteModelManager = model
+    init(model: NoteViewModel) {
+        self.viewModel = model
```
argument label도 `viewModel`로 동일하게 가져가는게 좋아보입니다.

hayeonhee replied:
이 부분 `viewModel`로 동일하게 변경했습니다! 😄

protocorn93 commented on CloudNotes/CloudNotes/Controller/NoteListTableViewDiffableDataSource.swift:
```diff
@@ -0,0 +1,36 @@
+    
+    init(
+        model: NoteViewModel,
+        tableView: UITableView,
```
`NoteListTableViewDiffableDataSource`가 `viewModel`을 프로퍼티로 가지고 있을 필요는 없어보입니다. 스와이프 액션에 대한 클로저만 전달해주는 것이 더 바람직해보여요. 

`viewModel`과의 강결합은 불필요해보입니다.

hayeonhee replied:
맞습니다. 
이 부분 스와이프 액션에 대한 코드 변화로 인해 `NoteListTableViewDiffableDataSource`가 `viewModel`을 알지 않아도 되어 해당 부분 삭제되었습니다.

protocorn93 commented on CloudNotes/CloudNotes/Model/Note.swift:
```diff
@@ -1,9 +1,22 @@
+            title: cdNote.title ?? "",
+            body: cdNote.body ?? "",
+            lastModified: cdNote.lastModified)
+    }
```
이렇게 `static func convertToNote(from:)` 메서드를 통해 `Note` 인스턴스를 만드는 방법도 있지만  `Note` 구조체에 `CDNote`를 인자로 받는 생성자를 만드는 방법도 있는데, 전자를 선택하신 이유가 궁금하네요~

smart8612 replied:
struct 의 멤버와이즈 생성자를 살리기 위해 타입 메서드 방식으로 구현하였습니다. 멤버와이즈 생성자는 sturct의 default 생성자인데, 사용자가 해당 타입의 custom 생성자를 구현하면 멤버와이즈 생성자를 사용할 수 없기 때문입니다.

다만 extension을 통해 custom 생성자를 구현했을 때 맴버와이즈 생성자가 살아있음을 확인할 수 있었기에 수정 구현하였습니다.

다음은 스위프트 프로그래밍 가이드에서 발췌한 내용입니다.
![unknown](https://user-images.githubusercontent.com/25794814/155293320-f5932052-3c26-44c3-bd37-860f63f97c1c.png)

protocorn93 reviewed (changes requested):
세 분 모두 고생하셨습니다! 쉽지 않은 프로젝트임에도 훌륭하게 진행해주셨어요👏

코드에 대한 코멘트 이외에 구조적인 코멘트는 하단에 "데이터 바인딩 역할 분리 관련" 섹션에서 추가로 남겨놓겠습니다. 



# 고민했던 점 

## 프로젝트 설계 구현 패턴 (MVVM)

### 기존의 `NoteModelManager`를 `NoteModel`과 `NoteViewModel`로 분리 구현하였습니다.

> - A 화면의 ViewController 입장에서는 Model에 불필요한 API 가 많이 보이게 됩니다.

단순히 동일한 일을 한다고 재사용한다는 것은 매우 위험합니다! 이렇게 불필요한 인터페이스가 노출된다고 생각이 든다면 분리를 해보셔야 합니다. 



### 프로젝트 전체에서 데이터가 단방향 흐름을 갖도록 설계하였습니다.

> - 다시 말해, 관리 해줘야 하는 데이터가 많기 때문에 데이터의 양에 비례하여 Controller 객체 구현이 복잡해질 수 있습니다.

맞아요! 컨트롤러의 코드가 복잡해진다는 뜻은 결국 데이터, 값의 흐름을 파악하기 힘들어지며 이는 곳 디버깅의 어려움으로 이어집니다. 그렇게 되면 유지보수성도 함께 저하되겠죠! 



### `UIViewController`의 새로운 관점

> 시스템 이벤트와 사용자의 터치 이벤트 등을 제어할 수 있는 View 라는 관점으로 보게 된 것 같습니다.

맞아요! 저도 뷰컨트롤러를 이런 관점으로 바라봅니다! 뷰의 레이아웃과 뷰컨트롤러 라이프사이클 이벤트 전달을 담당하는 동시에 사용자 이벤트 전달 등을 담당하는 뷰로 바라봅니다. 비즈니스 로직은 다른 객체가 담당하도록 하는 것이죠! 훌륭합니다! 👏



### Model 과 ViewController 간의 강한 결합성이 잔류하고 있습니다.

어디까지 추상화할 것인지는 오로지 본인과 팀의 판단입니다. 추상화를 하면 할 수록 코드는 보기 어려워집니다. 대신 유연성이나 확장성을 얻을 수 있죠. 가독성도 무시하지 못할 요소입니다. 

그렇기 때문에 저는 모델까지는 추상화를 진행하지 않아요. 대신 모델을 계층별로 나눕니다. 

- 뷰에서 화면에 값을 출력하는데 사용되는 모델 
- 비즈니스 로직에서 사용되는 모델 
- 외부 저장소(서버, 로컬DB)에서 가져온 값과 매핑되는 모델 

이렇게 나누고 각 계층에서 이들을 사용하는 객체에서 계층간 상호작용시 값을 적절히 변환해주곤 합니다. 

---



# 해결이 되지 않은 점

### `textViewDidChange()` 관련

이 이슈는 타이핑을 한번 할 때마다 로컬 DB 저장이라는 상대적으로 리소스가 많이 들어가는 작업이기 때문에 그렇게 보여집니다. 

이와 비슷한 예를 한번 생각해보면 앱내 검색창에서 원하는 키워드를 타이핑할 때마다 api 호출을 하면 리소스 낭비가 일어날 수 있어요. 그렇기 때문에 보통 이러한 기능은 두 가지 방법으로 리소스 이슈를 해결할 수 있어요. 

- 실시간으로 api를 호출하지말고 타이핑이 완료된 후 `검색`이라는 버튼을 눌러 api를 요청한다. (사용성이 저해될 수 있습니다.)
- throttle 기법을 통해 마지막 타이핑이 끝난 후 미세한 시간 텀을 두어 api를 호출한다. 

저는 보통 두 번째 방법을 많이 사용합니다. 이 예를 노트앱으로 가져와서 생각해보면 첫 번째 방법은 왼쪽 사이드메뉴에서 다른 메뉴나 기타 다른 버튼 이벤트가 발생했을 때 현재 화면을 벗어나기 전에 저장을 하는 것입니다. 

두번째 방법으로는 `Timer`를 사용해서 사용자가 입력을 마치고 일정 시간이 지났을 때 저장을 하고, 일정 시간 내에서 다시 타이핑이 일어나면 타이머는 초기화되어야겠죠! (나중에 여러분들도 경험하게 되실 수 있지만 RxSwift라는 리액티브 스트림 기반의 코드를 작성할 수 있는 라이브러리에서 자주 사용되는 기법입니다.)

이런 방법들로 앱 동작이 원할하지 않은 이슈를 해결해볼 수 있을 것 같아요.



### Cancel 액션 구현 관련 

이 부분은 현재 상태를 유지해도 상관없어 보입니다. 



### UIActivityViewController의 Options 버튼 관련

이 부분의 나머지 영역은 `LPLinkMetadata`로 커스터마이징이 가능해보이는데, `options`에 대해서는 저도 잘 알지 못하네요. 저도 한번 알아보고 추후에 공유해보겠습니다.

---



# 조언을 얻고 싶은 부분

 

### lazy var 관련

> Lazy 키워드를 사용하여 `NoteDetailViewController` 인스턴스 생성 후 필요할 때 bar button item 을 생성하여 target을 등록하면 target action이 잘 동작할 것이다.

네 맞습니다. 이게 원인이고 해결방법도 적절하게 잘 사용하셨습니다! 



### 데이터 바인딩 역할 분리 관련

`NSFetchedResultsControllerDelegate`는 여러분이 작성한 프로젝트 내에서 `NoteModel` 혹은 `PersistentDataManager`가 채택해야 한다고 생각해요. 구조적인 이야기를 해보자면 뷰모델을 그리는 방법은 다양합니다. 뷰모델의 형태가 다양한 것이죠. 그중에서 저는 다음과 같은 형태의 뷰모델을 선호합니다. 

*<u>뷰모델은 뷰의 추상화다. 뷰로부터 들어오는 이벤트를 적절한 비즈니스 로직 객체에게 전달하고 결과를 받아 뷰가 사용하는 형태로의 가공만을 담당한다.</u>*

저는 뷰모델을 이렇게 바라봅니다. 그렇기 때문에 뷰모델은 비즈니스 로직 객체가 데이터를 CoreData를 통해 로컬에 저장하던 Realm을 통해 로컬에 저장하던 서버에 저장하던 상관없고 모릅니다. 단지 뷰에서 "저장"이라는 버튼이 눌렸다는 이벤트를 데이터 저장이라는 "비즈니스 로직을 수행하는 객체"에 전달할 뿐입니다. 그리고 그 결과를 받아 다시 화면에 뿌려주는 것이죠. (VC ↔️ VM ↔️ Service )

이런 관점으로 코드를 구조를 다시 생각해보시면 좋을 것 같아요. 뷰모델이 어디에 데이터를 저장하는지 모르고 저장 이벤트를 전달하고 결과를 받는 구조라면 Mock 객체로 보다 쉽게 뷰모델을 테스트할 수 있습니다. 우리가 뷰모델에게 기대하는 바는 이벤트를 전달하고 결과를 받아 적절한 형태로 가공해서 돌려주는것이기 때문입니다. 

자 여기서 추상화라는 개념이 등장합니다. 우리는 추상화를 진행할 때 무엇을 추상화해야할까에 대한 고민을 자주 하곤합니다. 여기서 저만의 팁이라면 팁인데, 저는 이때 변하는 것과 변하지 않는 것을 구분합니다. 

현재 상황에서 변하지 않는 것은 데이터 관리라는 행위입니다. (CRUD) 반면, 변할 수 있는 것은 저장소입니다. (CoreData, Realm, Server 등등) 그렇다면 우리는 CRUD라는 행위를 추상화해야겠죠. 그리고 실제 매체를 다루는 객체가 이 추상화를 구체화하는겁니다. 

CRUD에서 CRU만 지원할 수 있는 것 아닌가요? 라는 의문이 생길 수 있어요. 이는 비즈니스 로직 자체가 변경이 되는 것이기에 아예 다른 비즈니스 로직이라고 생각하시면 되요. 이런것을 우리는 추상화 과정에서 변하는 것으로 구분하진 않습니다. 

현재 여러분들이 사용하는 구조는 MVVM에서 Repository를 사용하는 것과 유사한 구조를 가집니다. 이것과 관련해서 여러 자료들을 찾아보시면 인사이트를 얻으실 수 있을 것 같아요

hayeonhee committed:
feat: swipe action 구현 #6

hayeonhee committed:
refactor: 불필요한 코드와 개행 삭제

hayeonhee committed:
fix: swipe action 구현 및 적용 #6

smart8612 committed:
fix: 노트 생성시 발생하는 오류 수정 #6

- 새로운 노트 생성시 detail view에 노트 세부 내용을 조회할 수 없는 문제점 해결
- fetch result controller delegate 함수 선언부 코딩 컨벤션 개선

Kim-EunsooSilver committed:
refactor: 접근제어 설정, 변수 -> 상수로 변경, 네이밍 변경 #7

hayeonhee committed:
refactor: 네이밍 수정, 메서드와 이니셜라이저 생성 및 삭제 #7

- changeTextViewEditingByIdentifier, resetDetailTextView 메서드 생성
- Section enum을 NoteTableViewController 타입 내부로 이동
- NoteTableViewController 네이밍 수정
- Note의 convertToNote static 메서드를 이니셜라이저로 변경
- NoteTableViewDiffableDataSource 내 사용하지 않는 메서드 삭제
